File: pipesmanufacturing/pipes_manufacturing/utils/material_request.py
from __future__ import unicode_literals
import frappe
from frappe.model.document import Document

@frappe.whitelist()
def raise_pipes_work_orders(material_request):
    mr= frappe.get_doc("Material Request", material_request)
    pipes_orders = []
    default_wip_warehouse = frappe.db.get_single_value("Manufacturing Settings", "default_wip_warehouse")
    for d in mr.items:
        if (d.qty - d.ordered_qty) >0:
            pipes_order = frappe.new_doc("Pipes Work Order")
            pipes_order.update({
                "production_item": d.item_code,
                "wip_warehouse": default_wip_warehouse,
                "t_warehouse": d.warehouse,
                "planned_start_date": mr.transaction_date,
                "planned_end_date": d.schedule_date,
                "material_request": mr.name,
                "req_qty": d.qty - d.ordered_qty,
            })
            pipes_order.save()
            mr.db_set("pipes_work_order",pipes_order.name)
            #pipes_orders.append(pipes_order.name)
            frappe.msgprint("Pipes Work Order# " + pipes_order.name + " created")


def verify_items(self,frm):
        item_cmp1 = None
        item_cmp2 = None
        throw_error = 0
        for i in range(len(self.items)):
            item_cmp1 = self.items[len(self.items)-i-1].item_code
            for j in range(len(self.items)):
                
                if j < len(self.items)-i-1:
                    item_cmp2 = self.items[j].item_code
                    if item_cmp1 == item_cmp2:
                        #frappe.msgprint('{0} is alreaded added in row {1}'.format(item_cmp1,j+1))
                        throw_error = 1
        has_non_pipe = 0
        has_pipe = 0
        for i in self.items:
            if 'Pipe-MS' in str(i.item_code):
                has_pipe = 1
            else:
                has_non_pipe = 1
        if has_pipe == has_non_pipe == 1:
            frappe.throw('Cannot add Pipe and Non-Pipe items together please correct!')
        if throw_error == 1:
            frappe.throw('Cannot add an item multiple times please correct!')

File: pipesmanufacturing/pipes_manufacturing/utils/material_request.js
var refresh_check = 0;
frappe.ui.form.on('Material Request', {
    refresh: function(frm) {
		if (refresh_check == 0){
			frm.events.make_custom_buttons(frm);
		}
		
	},
	
	onload: function(){
		refresh_check = 0;
	},

    make_custom_buttons: function(frm) {
        if (frm.doc.docstatus == 1 && frm.doc.status != 'Stopped' ) {
			refresh_check = 1;
			var has_pipe = 0;
			var has_non_pipe = 0;
			var make_pwo_button = 0;  
			for (var i in cur_frm.doc.items){
				if (cur_frm.doc.items[i].item_code.includes("Pipe-MS",0)){
					has_pipe = 1;
					if (cur_frm.doc.items[i].ordered_qty!=cur_frm.doc.items[i].qty){
						make_pwo_button = 1;
					}
				}
				else{
					has_non_pipe = 1;
				}
			}
            if (flt(frm.doc.per_ordered, 2) < 100) {
                // make
                if (frm.doc.material_request_type === "Manufacture" && has_pipe == 1 && make_pwo_button == 1 && cur_frm.doc.pipes_work_order == undefined) {
					frm.add_custom_button(__("Pipes Work Order"),
						() => frm.events.raise_pipes_work_orders(frm), __('Create'));
				}
				if (has_non_pipe == 1){
					frm.add_custom_button(__("Work Order"),
					() => frm.events.raise_work_orders(frm), __('Create'));
				}
            }
        }
	},
	
    raise_pipes_work_orders: function(frm) {
		frappe.call({
			method:"pipesmanufacturing.pipes_manufacturing.utils.material_request.raise_pipes_work_orders",
			args: {
				"material_request": frm.doc.name
			},
			callback: function(r) {
				if(r) {
					frm.reload_doc();
				}
			}
		});
	},

	raise_work_orders: function(frm) {
		frappe.call({
			method:"erpnext.stock.doctype.material_request.material_request.raise_work_orders",
			args: {
				"material_request": frm.doc.name
			},
			callback: function(r) {
				if(r.message.length) {
					frm.reload_doc();
				}
			}
		});
	}
})
/*
frappe.ui.form.on("Material Request Item",{
	item_code: function(frm,cdt,cdn){
		var item = frappe.model.get_doc(cdt,cdn);
		if (cur_frm.doc.items.length!=1){
			for (var i=0; i< cur_frm.doc.items.length-1; i++){
				if (cur_frm.doc.items[i].item_code == item.item_code){
					frappe.throw("Item " + item.item_code + " is already added!")
				}
			}
		}
		
	}
})
*/